src/client/pythonEnvironments/creation/proposed.createEnvApis.ts

```diff
@@ -0,0 +1,161 @@
+     */
+    showBackButton?: boolean;
+
+    /**
+     * Default `true`. If `true`, the environment will be selected as the environment to be used for the workspace.
+     */
+    selectEnvironment?: boolean;
+}
+
+/**
+ * Params passed on `onWillCreateEnvironment` event handler.
+ */
+export interface WillCreateEnvironmentParams {
+    /**
+     * Options used to create a Python environment.
+     */
+    options: CreateEnvironmentOptions | undefined;
+}
+
+/**
+ * Params passed on `onDidCreateEnvironment` event handler.
+ */
+export interface DidCreateEnvironmentParams {
+    /**
+     * Options used to create the Python environment.
+     */
+    options?: CreateEnvironmentOptions;
+
+    /**
+     * Workspace Uri associated with the environment.
+     */
+    uri?: Uri;
+
+    /**
+     * Path to the executable python in the environment
+     */
+    path?: string;
+
+    /**
+     * User action that resulted in exit from the create environment flow.
+     */
+    action?: CreateEnvironmentUserActions;
+
+    /**
+     * Error if any occurred during environment creation.
+     */
+    error?: Error;
+}
+
+export interface CreateEnvironmentResult {
+    /**
+     * Workspace Uri associated with the environment.
+     */
+    uri?: Uri;
+
+    /**
+     * Path to the python executable in the environment.
+     */
+    path?: string;
+
+    /**
+     * User action that resulted in exit from the create environment flow.
+     */
+    action?: CreateEnvironmentUserActions;
+
+    /**
+     * Error if any occurred during environment creation.
+     */
+    error?: Error;
+}
+
+/**
+ * Extensions that want to contribute their own environment creation can do that by registering an object
+ * that implements this interface.
+ */
+export interface CreateEnvironmentProvider {
+    /**
+     * This API is called when user selects this provider from a QuickPick to select the type of environment
+     * user wants. This API is expected to show a QuickPick or QuickInput to get the user input and return
+     * the path to the Python executable in the environment.
+     *
+     * @param {CreateEnvironmentOptions} [options] Options used to create a Python environment.
+     *
+     * @returns a promise that resolves to the path to the
+     * Python executable in the environment. Or any action taken by the user, such as back or cancel.
+     */
+    createEnvironment(options?: CreateEnvironmentOptions): Promise<CreateEnvironmentResult | undefined>;
+
+    /**
+     * Unique ID for the creation provider, typically <ExtensionId>:<environment-type | guid>
+     */
+    id: CreateEnvironmentProviderId;
+
+    /**
+     * Display name for the creation provider.
+     */
+    name: string;
+
+    /**
+     * Description displayed to the user in the QuickPick to select environment provider.
+     */
+    description: string;
+
+    /**
+     * Tools used to manage this environment. e.g., ['conda']
+     */
+    tools?: string[];
```

Comment:
```suggestion
    /**
     * Primary tool used to create the environment. e.g., ['conda']
     */
    tool: EnvironmentTools;
```

I would expect it only contain the one tool, so it can be used instead of `id`.

Reply:
What will happen if an extension wants to register multiple ways to create environment with the same tool? `id` is never a substitute for tool names. What if extension want to make an environment usable by multiple tools, like conda and mamba?  The common case is there is one tool, but we should program for extensibility.

Reply:
I agree, it cannot be a substitute to `id`.

> What if extension want to make an environment usable by multiple tools, like conda and mamba?

I'm not clear about this. There is only one tool which manages an environment, it's the one which created it. It uses either mamba CLI or conda CLI, otherwise I think we're enforcing bad practice.

If it has to be an array, we should make it clear that the primary tool comes first in the list so it's prioritized.

Reply:
> otherwise I think we're enforcing bad practice.

There are tools that wrap conda environments into portable environments. Which are used as standard way of working with environments. We should not block them. I will make this a prioritized list.